bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/remote/ApiClientExt.kt

```diff
@@ -0,0 +1,75 @@
+package io.bucketeer.sdk.android.internal.remote
+
+import com.squareup.moshi.JsonAdapter
+import io.bucketeer.sdk.android.BKTException
+import io.bucketeer.sdk.android.internal.model.response.ErrorResponse
+import okhttp3.Response
+import kotlin.contracts.ExperimentalContracts
+
+fun Response.toBKTException(adapter: JsonAdapter<ErrorResponse>): BKTException {
+
+  val bodyString = body?.string() ?: ""
+  val errorBody: ErrorResponse? = kotlin.runCatching {
+    adapter.fromJson(bodyString)
+  }.getOrNull()
+
+  return when (code) {
+    400 -> {
+      // BadRequest
+      // - gateway: context canceled
+      // - gateway: tag is required
+      // - gateway: user is required
+      // - gateway: user id is required
+      // - gateway: feature id is required
+      // - gateway: missing event id
+      // - gateway: missing events
+      // - gateway: body is required
+      BKTException.BadRequestException(
+        message = errorBody?.error?.message ?: "BadRequest error"
+      )
+    }
+    401 -> {
+      // Unauthorized
+      // - gateway: missing APIKey
+      // - gateway: invalid APIKey
+      // - gateway: bad role
+      // - gateway: disabled APIKey
+      BKTException.UnauthorizedException(
+        message = errorBody?.error?.message ?: "Unauthorized error"
+      )
+    }
+    404 -> {
+      // NotFound
+      // - feature not found
+      BKTException.FeatureNotFoundException(
+        message = errorBody?.error?.message ?: "NotFound error"
+      )
+    }
+    405 -> {
+      // MethodNotAllowed
+      // - gateway: invalid http method
+      BKTException.InvalidHttpMethodException(
+        message = errorBody?.error?.message ?: "MethodNotAllowed error"
+      )
+    }
+    500 -> {
+      // InternalServerError
+      // - gateway: internal
+      BKTException.ApiServerException(
+        message = errorBody?.error?.message ?: "InternalServer error"
+      )
+    }
+    else -> BKTException.UnknownException("Unknown error: '$errorBody'")
```

Comment:
Just confirming, would it be possible also to handle network errors? Such as no connection and timeout?
I'm asking because I want to report network errors via metrics events.

Reply:
Yes, I'm going to add `TimeoutException` and `NetworkException`

Reply:
added [`de0d31b` (#62)](https://github.com/ca-dp/bucketeer-android-sdk/pull/62/commits/de0d31bf2bb21becc39580b3c21b334ef3f468c3)